README
Welcome
-------

The 0MQ lightweight messaging kernel is a library which extends the
standard socket interfaces with features traditionally provided by
specialised messaging middleware products. 0MQ sockets provide an
abstraction of asynchronous message queues, multiple messaging patterns,
message filtering (subscriptions), seamless access to multiple transport
protocols and more.

Atomic zeromq modifies some of the messaging guarantees that zeromq makes
to allow differentiated publisher policies.
The modified library allows one to set messaging policy on sockets such
as best effort/ all or none and so on.


Building and installation
-------------------------

See the INSTALL file included with the distribution.


Resources
---------

Extensive documentation is provided with the distribution. Refer to
doc/zmq.html, or "man zmq" after you have installed 0MQ on your system.

Website: http://www.zeromq.org/

This is a modified version of zeromq which uses an atomic variation of publish subscribe.
This means that the messages will not be silently dropped instead zeromq will accept a sendmsg
only when the message can be delivered to all eligible recipients.



Git repository: http://github.com/zeromq/libzmq

0MQ developers can also be found on the IRC channel #zeromq, on the
Freenode network (irc.freenode.net).

Copying
-------

Free use of this software is granted under the terms of the GNU Lesser General
Public License (LGPL). For details see the files `COPYING` and `COPYING.LESSER`
included with the 0MQ distribution.


ChangeLog
# Generated by Makefile. Do not edit.

commit 3bf7860fb783c38503c013dbad05ad2ff54f5a16
Author: Martin Sustrik <sustrik@250bpm.com>
Date:   Sun Dec 18 11:41:45 2011 +0100

    NEWS updated for 3.1.0 release
    
    Signed-off-by: Martin Sustrik <sustrik@250bpm.com>

 NEWS |  637 ++++++++++--------------------------------------------------------
 1 files changed, 90 insertions(+), 547 deletions(-)

commit 22ef966d4aab3448d4debf524afbdaa19f37dff0
Author: Martin Lucina <martin@lucina.net>
Date:   Sun Dec 18 11:19:55 2011 +0100

    Update email address in man pages
    
    Signed-off-by: Martin Lucina <martin@lucina.net>

 doc/zmq.txt               |    2 +-
 doc/zmq_bind.txt          |    2 +-
 doc/zmq_close.txt         |    2 +-
 doc/zmq_connect.txt       |    2 +-
 doc/zmq_errno.txt         |    2 +-
 doc/zmq_getsockopt.txt    |    2 +-
 doc/zmq_init.txt          |    2 +-
 doc/zmq_inproc.txt        |    2 +-
 doc/zmq_ipc.txt           |    2 +-
 doc/zmq_msg_close.txt     |    2 +-
 doc/zmq_msg_copy.txt      |    2 +-
 doc/zmq_msg_data.txt      |    2 +-
 doc/zmq_msg_init.txt      |    2 +-
 doc/zmq_msg_init_data.txt |    2 +-
 doc/zmq_msg_init_size.txt |    2 +-
 doc/zmq_msg_move.txt      |    2 +-
 doc/zmq_msg_size.txt      |    2 +-
 doc/zmq_pgm.txt           |    2 +-
 doc/zmq_poll.txt          |    2 +-
 doc/zmq_recv.txt          |    2 +-
 doc/zmq_recvmsg.txt       |    2 +-
 doc/zmq_send.txt          |    2 +-
 doc/zmq_sendmsg.txt       |    2 +-
 doc/zmq_setsockopt.txt    |    2 +-
 doc/zmq_socket.txt        |    2 +-
 doc/zmq_strerror.txt      |    2 +-
 doc/zmq_tcp.txt           |    2 +-
 doc/zmq_term.txt          |    2 +-
 doc/zmq_version.txt       |    2 +-
 29 files changed, 29 insertions(+), 29 deletions(-)

commit 183e126364762eb7b6fcc96c87d060c8e09044b5
Author: Martin Lucina <martin@lucina.net>
Date:   Sun Dec 18 11:12:44 2011 +0100

    Fix typo in zmq_sendmsg(3) manual page
    
    Signed-off-by: Martin Lucina <martin@lucina.net>

 doc/zmq_sendmsg.txt |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 3fbe8ac3c3aaec777c305de8fbdc0d748c3b14aa
Author: Martin Sustrik <sustrik@250bpm.com>
Date:   Sat Dec 17 23:42:45 2011 +0100

    Dont wait for pending subscription commands on shutdown
    
    When (X)SUB socket is being shut down there may be pending
    outbound subscription commands in the pipes. In such case
    we want to close the socket immediately instead of waiting
    for the commands to be sent.
    
    Signed-off-by: Martin Sustrik <sustrik@250bpm.com>

 src/xsub.cpp |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

commit 91fdedf25c4d76b0ec0aeb5d1d9f1c9a1a769447
Author: Martin Sustrik <sustrik@250bpm.com>
Date:   Sat Dec 17 10:14:32 2011 +0100

    Fix polling on XREP socket
    
    When polling on XREP socket in incoming message part was prefetched,
    but not the identity of sender. The problem is fixed by this patch.
    
    Signed-off-by: Martin Sustrik <sustrik@250bpm.com>

 src/xrep.cpp |   45 ++++++++++++++++++++++++++++++++++++---------
 src/xrep.hpp |    8 ++++++--
 2 files changed, 42 insertions(+), 11 deletions(-)

commit f9eb763293014f812dac5558be5c5f03bb896efb
Author: Martin Sustrik <sustrik@250bpm.com>
Date:   Fri Dec 16 07:04:38 2011 +0100

    zmq_sendmsg and zmq_recvmsg checks for NULL message object
    
    Signed-off-by: Martin Sustrik <sustrik@250bpm.com>

 src/socket_base.cpp |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

commit 3fb5c11b03119a157db8dc024809943f88f87320
Author: Stuart Webster <sw_webster@hotmail.com>
Date:   Thu Dec 15 14:03:43 2011 +0100

    Added MSVC10 solution and project files for Windows
    
    These include configurations for both Win32 and x64 platforms. All project
    settings have been normalised in property sheets (the ".props" files under
    builds/msvc/properties) to simplify maintenance. Build artefacts are all
    generated in platform-specific subfolders of bin, lib and obj directories.
    
    Also enables the use of precompiled headers with MSVC10.
    
    This significantly reduces the time required to compile libzmq with Visual
    Studio on Windows. It should have no impact on other platforms.
    
    Signed-off-by: Stuart Webster <sw_webster@hotmail.com>

 .gitignore                                    |    1 +
 builds/msvc/c_local_lat/c_local_lat.vcxproj   |   87 +++++++++
 builds/msvc/c_local_thr/c_local_thr.vcxproj   |   87 +++++++++
 builds/msvc/c_remote_lat/c_remote_lat.vcxproj |   87 +++++++++
 builds/msvc/c_remote_thr/c_remote_thr.vcxproj |   87 +++++++++
 builds/msvc/inproc_lat/inproc_lat.vcxproj     |   86 +++++++++
 builds/msvc/inproc_thr/inproc_thr.vcxproj     |   86 +++++++++
 builds/msvc/libzmq/libzmq.vcxproj             |  241 +++++++++++++++++++++++++
 builds/msvc/libzmq/libzmq.vcxproj.filters     |   13 ++
 builds/msvc/msvc10.sln                        |  116 ++++++++++++
 builds/msvc/properties/Common.props           |   21 +++
 builds/msvc/properties/Debug.props            |   19 ++
 builds/msvc/properties/Dynamic.props          |   20 ++
 builds/msvc/properties/Executable.props       |   19 ++
 builds/msvc/properties/Precompiled.props      |   14 ++
 builds/msvc/properties/Release.props          |   22 +++
 builds/msvc/properties/Win32.props            |   12 ++
 builds/msvc/properties/Win32_Release.props    |   17 ++
 builds/msvc/properties/WithOpenPGM.props      |   12 ++
 builds/msvc/properties/ZeroMQ.props           |   23 +++
 builds/msvc/properties/x64.props              |   12 ++
 src/precompiled.cpp                           |   21 +++
 src/precompiled.hpp                           |   47 +++++
 src/windows.hpp                               |    3 +
 24 files changed, 1153 insertions(+), 0 deletions(-)

commit a9e03336b020d2e6a83338b111092e65aa812f95
Author: Martin Lucina <martin@lucina.net>
Date:   Thu Dec 15 13:23:40 2011 +0100

    A synchronous connect() failure in ipc_connecter can result in Assertion

include/zmq.h
/*
    Copyright (c) 2009-2011 250bpm s.r.o.
    Copyright (c) 2007-2010 iMatix Corporation
    Copyright (c) 2011 VMware, Inc.
    Copyright (c) 2007-2011 Other contributors as noted in the AUTHORS file

    This file is part of 0MQ.

    0MQ is free software; you can redistribute it and/or modify it under
    the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    0MQ is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef __ZMQ_H_INCLUDED__
#define __ZMQ_H_INCLUDED__
/*
#ifdef __cplusplus
extern "C" {
#endif
*/
#include <vector>
#include <string>

#include <errno.h>
#include <stddef.h>
#if defined _WIN32
#include <winsock2.h>
#endif

/*  Handle DSO symbol visibility                                             */
#if defined _WIN32
#   if defined DLL_EXPORT
#       define ZMQ_EXPORT __declspec(dllexport)
#   else
#       define ZMQ_EXPORT __declspec(dllimport)
#   endif
#else
#   if defined __SUNPRO_C  || defined __SUNPRO_CC
#       define ZMQ_EXPORT __global
#   elif (defined __GNUC__ && __GNUC__ >= 4) || defined __INTEL_COMPILER
#       define ZMQ_EXPORT __attribute__ ((visibility("default")))
#   else
#       define ZMQ_EXPORT
#   endif
#endif

/******************************************************************************/
/*  0MQ versioning support.                                                   */
/******************************************************************************/

/*  Version macros for compile-time API version detection                     */
#define ZMQ_VERSION_MAJOR 3
#define ZMQ_VERSION_MINOR 1
#define ZMQ_VERSION_PATCH 0

#define ZMQ_MAKE_VERSION(major, minor, patch) \
    ((major) * 10000 + (minor) * 100 + (patch))
#define ZMQ_VERSION \
    ZMQ_MAKE_VERSION(ZMQ_VERSION_MAJOR, ZMQ_VERSION_MINOR, ZMQ_VERSION_PATCH)

/*  Run-time API version detection                                            */
ZMQ_EXPORT void zmq_version (int *major, int *minor, int *patch);

/******************************************************************************/
/*  0MQ errors.                                                               */
/******************************************************************************/

/*  A number random enough not to collide with different errno ranges on      */
/*  different OSes. The assumption is that error_t is at least 32-bit type.   */
#define ZMQ_HAUSNUMERO 156384712

/*  On Windows platform some of the standard POSIX errnos are not defined.    */
#ifndef ENOTSUP
#define ENOTSUP (ZMQ_HAUSNUMERO + 1)
#endif
#ifndef EPROTONOSUPPORT
#define EPROTONOSUPPORT (ZMQ_HAUSNUMERO + 2)
#endif
#ifndef ENOBUFS
#define ENOBUFS (ZMQ_HAUSNUMERO + 3)
#endif
#ifndef ENETDOWN
#define ENETDOWN (ZMQ_HAUSNUMERO + 4)
#endif
#ifndef EADDRINUSE
#define EADDRINUSE (ZMQ_HAUSNUMERO + 5)
#endif
#ifndef EADDRNOTAVAIL
#define EADDRNOTAVAIL (ZMQ_HAUSNUMERO + 6)
#endif
#ifndef ECONNREFUSED
#define ECONNREFUSED (ZMQ_HAUSNUMERO + 7)
#endif
#ifndef EINPROGRESS
#define EINPROGRESS (ZMQ_HAUSNUMERO + 8)
#endif
#ifndef ENOTSOCK
#define ENOTSOCK (ZMQ_HAUSNUMERO + 9)
#endif
#ifndef EAFNOSUPPORT
#define EAFNOSUPPORT (ZMQ_HAUSNUMERO + 10)
#endif

/*  Native 0MQ error codes.                                                   */
#define EFSM (ZMQ_HAUSNUMERO + 51)
#define ENOCOMPATPROTO (ZMQ_HAUSNUMERO + 52)
#define ETERM (ZMQ_HAUSNUMERO + 53)
#define EMTHREAD (ZMQ_HAUSNUMERO + 54)

/*  This function retrieves the errno as it is known to 0MQ library. The goal */
/*  of this function is to make the code 100% portable, including where 0MQ   */
/*  compiled with certain CRT library (on Windows) is linked to an            */
/*  application that uses different CRT library.                              */
ZMQ_EXPORT int zmq_errno (void);

/*  Resolves system errors and 0MQ errors to human-readable string.           */
ZMQ_EXPORT const char *zmq_strerror (int errnum);

/******************************************************************************/
/*  0MQ message definition.                                                   */
/******************************************************************************/

typedef struct {unsigned char _ [32];} zmq_msg_t;

typedef void (zmq_free_fn) (void *data, void *hint);

ZMQ_EXPORT int zmq_msg_init (zmq_msg_t *msg);
ZMQ_EXPORT int zmq_msg_init_size (zmq_msg_t *msg, size_t size);
ZMQ_EXPORT int zmq_msg_init_data (zmq_msg_t *msg, void *data,
    size_t size, zmq_free_fn *ffn, void *hint);
ZMQ_EXPORT int zmq_msg_close (zmq_msg_t *msg);
ZMQ_EXPORT int zmq_msg_move (zmq_msg_t *dest, zmq_msg_t *src);
ZMQ_EXPORT int zmq_msg_copy (zmq_msg_t *dest, zmq_msg_t *src);
ZMQ_EXPORT void *zmq_msg_data (zmq_msg_t *msg);
ZMQ_EXPORT size_t zmq_msg_size (zmq_msg_t *msg);
ZMQ_EXPORT int zmq_getmsgopt (zmq_msg_t *msg, int option, void *optval,
    size_t *optvallen);

/******************************************************************************/
/*  0MQ infrastructure (a.k.a. context) initialisation & termination.         */
/******************************************************************************/

ZMQ_EXPORT void *zmq_init (int io_threads);
ZMQ_EXPORT int zmq_term (void *context);

/******************************************************************************/
/*  0MQ socket definition.                                                    */
/******************************************************************************/

/*  Socket types.                                                             */ 
#define ZMQ_PAIR 0
#define ZMQ_PUB 1
#define ZMQ_SUB 2
#define ZMQ_REQ 3
#define ZMQ_REP 4
#define ZMQ_XREQ 5
#define ZMQ_XREP 6
#define ZMQ_PULL 7
#define ZMQ_PUSH 8
#define ZMQ_XPUB 9
#define ZMQ_XSUB 10

#define ZMQ_ROUTER ZMQ_XREP
#define ZMQ_DEALER ZMQ_XREQ

/*  Socket options.                                                           */
#define ZMQ_AFFINITY 4
#define ZMQ_IDENTITY 5
#define ZMQ_SUBSCRIBE 6
#define ZMQ_UNSUBSCRIBE 7
#define ZMQ_RATE 8
#define ZMQ_RECOVERY_IVL 9
#define ZMQ_SNDBUF 11
#define ZMQ_RCVBUF 12
#define ZMQ_RCVMORE 13
#define ZMQ_FD 14
#define ZMQ_EVENTS 15
#define ZMQ_TYPE 16
#define ZMQ_LINGER 17
#define ZMQ_RECONNECT_IVL 18
#define ZMQ_BACKLOG 19
#define ZMQ_RECONNECT_IVL_MAX 21
#define ZMQ_MAXMSGSIZE 22
#define ZMQ_SNDHWM 23
#define ZMQ_RCVHWM 24
#define ZMQ_MULTICAST_HOPS 25
#define ZMQ_RCVTIMEO 27
#define ZMQ_SNDTIMEO 28
#define ZMQ_IPV4ONLY 31

/*  Message options                                                           */
#define ZMQ_MORE 1

/*  Send/recv options.                                                        */
#define ZMQ_DONTWAIT 1
#define ZMQ_SNDMORE 2

ZMQ_EXPORT void *zmq_socket (void *context, int type);
ZMQ_EXPORT int zmq_close (void *s);
ZMQ_EXPORT int zmq_setsockopt (void *s, int option, const void *optval,
    size_t optvallen); 
ZMQ_EXPORT int zmq_getsockopt (void *s, int option, void *optval,
    size_t *optvallen);
ZMQ_EXPORT int zmq_bind (void *s, const char *addr);
ZMQ_EXPORT int zmq_connect (void *s, const char *addr);
ZMQ_EXPORT int zmq_send (void *s, const void *buf, size_t len, int flags);
ZMQ_EXPORT int zmq_recv (void *s, void *buf, size_t len, int flags);
ZMQ_EXPORT int zmq_sendmsg (void *s, zmq_msg_t *msg, int flags);
ZMQ_EXPORT int zmq_recvmsg (void *s, zmq_msg_t *msg, int flags);

typedef struct
{
    std::string identity;
    uint64_t msg_slot_free;
    int64_t buffer_space_avail;
} zmq_stats_t;

/* Stats reporting for publisher subscriber  */
ZMQ_EXPORT int zmq_get_stats(void* s, std::vector<zmq_stats_t>& stats  );




/******************************************************************************/
/*  I/O multiplexing.                                                         */
/******************************************************************************/

#define ZMQ_POLLIN 1
#define ZMQ_POLLOUT 2
#define ZMQ_POLLERR 4

typedef struct
{
    void *socket;
#if defined _WIN32
    SOCKET fd;
#else
    int fd;
#endif
    short events;
    short revents;
} zmq_pollitem_t;

ZMQ_EXPORT int zmq_poll (zmq_pollitem_t *items, int nitems, long timeout);

#undef ZMQ_EXPORT

/*
#ifdef __cplusplus
}
#endif
*/

#endif

